### PR TITLE
Fix vb event generation in navbars

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
We need integration tests for this.  That is tracked by: https://github.com/dotnet/roslyn/issues/53875

### CODE REVIEW

File: src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptNavigationBarItemService.cs [hunk 1]
```diff
@@ -40,16 +40,18 @@ internal class VSTypeScriptNavigationBarItemService : INavigationBarItemService
...
-            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                var workspace = document.Project.Solution.Workspace;
+                var navigationService = VSTypeScriptDocumentNavigationServiceWrapper.Create(workspace);
+                navigationService.TryNavigateToPosition(workspace, document.Id, item.Spans[0].Start, virtualSpace: 0, options: null, cancellationToken: cancellationToken);
+            }
 
-            var workspace = document.Project.Solution.Workspace;
-            var navigationService = VSTypeScriptDocumentNavigationServiceWrapper.Create(workspace);
-            navigationService.TryNavigateToPosition(workspace, document.Id, item.Spans[0].Start, virtualSpace: 0, options: null, cancellationToken: cancellationToken);
+            return true;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptNavigationBarItemService.cs:54]: note: i'm preserving behavior for TS and F#.  Neither of which have any sort of special 'type list item is selected, so update the member list' behavior like VB has.

File: src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs [hunk 2]
```diff
@@ -386,8 +386,14 @@ private async Task ProcessItemSelectionAsync(NavigationBarItem item, Cancellatio
...
-                    // ConfigureAwait(true) as we have to come back to UI thread in order to kick of the refresh task below.
-                    await navBarService.NavigateToItemAsync(document, item, view, cancellationToken).ConfigureAwait(true);
+                    // ConfigureAwait(true) as we have to come back to UI thread in order to kick of the refresh task
+                    // below. Note that we only want to refresh if selecting the item had an effect (either navigating
+                    // or generating).  If nothing happened to don't want to refresh.  This is important as some items
+                    // exist in the type list that are only there to show a set a particular set of items in the member
+                    // list.  So selecting such an item should only update the member list, and we do not want a refresh
+                    // to wipe that out.
+                    if (!await navBarService.TryNavigateToItemAsync(document, item, view, cancellationToken).ConfigureAwait(true))
+                        return;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs:396]: this regressed in:  https://github.com/dotnet/roslyn/pull/52530/files#diff-902dbb57c78c6bbd71f7409a96a3db1dbd4813fb5658a069ef3b3b33da063a8aL305

THe reason for this is that the old code effectively would say: ok, something was selected.  However the semantic version of the solution is the same, so i won't refresh anything.

The approach here is cleaner.  If trying to navigate failed, there's no need for us to have our controller change anything, and we'll instead just let the view update accordingly.

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicEditorNavigationBarItemService.vb [hunk 3]
```diff
@@ -73,16 +73,21 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
...
             Dim symbolItem = TryCast(underlying, SymbolItem)
             If generateCodeItem IsNot Nothing Then
                 Await GenerateCodeForItemAsync(document, generateCodeItem, textView, cancellationToken).ConfigureAwait(False)
+                Return True
             ElseIf symbolItem IsNot Nothing Then
                 Await NavigateToSymbolItemAsync(document, symbolItem, cancellationToken).ConfigureAwait(False)
+                Return True
             End If
+
+            Return False
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicEditorNavigationBarItemService.vb:90]: this is the critical part of the change.  We didn't navigate or generate, so this means the user effectively only picked the item just to get the refreshed navbars.

File: src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs [hunk 4]
```diff
@@ -386,8 +386,14 @@ private async Task ProcessItemSelectionAsync(NavigationBarItem item, Cancellatio
                     item.Spans = item.TrackingSpans.SelectAsArray(ts => ts.GetSpan(snapshot).Span.ToTextSpan());
                     var view = _presenter.TryGetCurrentView();
 
-                    // ConfigureAwait(true) as we have to come back to UI thread in order to kick of the refresh task below.
-                    await navBarService.NavigateToItemAsync(document, item, view, cancellationToken).ConfigureAwait(true);
+                    // ConfigureAwait(true) as we have to come back to UI thread in order to kick of the refresh task
+                    // below. Note that we only want to refresh if selecting the item had an effect (either navigating
+                    // or generating).  If nothing happened to don't want to refresh.  This is important as some items
+                    // exist in the type list that are only there to show a set a particular set of items in the member
```
Comment (davidwengier) [inline src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs:392]:
```suggestion
                    // exist in the type list that are only there to show a particular set of items in the member
```

**Review by davidwengier — APPROVED** [review]

File: src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs:392]: WIll do in next PR :)

**Pull request merged**